### ISSUE TITLE
#warning is not a valid preprocessor directive; code cannot compile on MSVC

### ISSUE THREAD

**Issue by codygray** [sciplot/sciplot#79]
#72 (merged by https://github.com/sciplot/sciplot/commit/49b3e2210682f5fcf17e71400a4b185076e2a6a0) introduced a bug with the `#warning` preprocessor directive.

This is not a valid preprocessor directive according to the C++ language standard. GCC supports it as an extension, but Microsoft's C++ compiler (as bundled with Visual Studio) does not. Because preprocessor directives that are not supported result in an *error*, rather than the compiler simply ignoring them (with a special exception made for `#pragma`), this prevents the code from being compiled with Microsoft's compiler (MSVC) at all. All you get is [error C1021](https://docs.microsoft.com/en-us/cpp/error-messages/compiler-errors-1/fatal-error-c1021):

> C1021: invalid preprocessor command 'warning'

On MSVC, you should be using `#pragma message` to generate a compile-time warning. (This is one of those rare cases where MSVC is actually doing the standards-complaint thing and using a `#pragma` for optional, implementation-specific features.)

The simplistic cross-platform solution is:

```c++
// We check if windows.h. is already included, as this might break compilation. See: https://github.com/sciplot/sciplot/issues/70
#ifdef _WINDOWS_
    #ifdef _MSC_VER
        #pragma message(__FILE__ "(): warning: You might run into compiler errors if windows.h is included before sciplot.hpp! See: https://sciplot.github.io/known_issues/")
    #else
        #warning You might run into compiler errors if windows.h is included before sciplot.hpp! See: https://sciplot.github.io/known_issues/
    #endif  // _MSC_VER
#endif  // _WINDOWS_
```

Note that the formatting used for `#pragma message` on MSVC allows the IDE to recognize the message as a warning, and double-clicking on it in the "Output" or "Error List" windows will even take you directly to the affected file. (Original credit for this trick goes to [James McNellis](https://stackoverflow.com/a/2706693).)

Technically, this isn't portable, either, as it unconditionally uses `#warning` on any compiler that is not MSVC. It should probably only be using `#warning` on any compiler that reports being GCC-compatible (*e.g.*, `#ifdef __GNUC__`), but support for `#warning` is more widespread than just GCC and those that emulate it. MSVC is the major "odd one out" here, so this is the minimal solution for the problem. If you want complete portability, look into doing what Hedley does, as described by its author here: https://stackoverflow.com/a/49469158

**Comment by HorstBaerbel** [comment]
Thanks for your contribution and showing me [Hedley](https://nemequ.github.io/hedley/).
Imo it is enough to support MSVC, GCC and CLANG for this project atm as this will probably cover >95% of the compilers used, so I'll merge your suggestion into master.

**Comment by HorstBaerbel** [comment]
Should no be fixed in da5db21
